src/inmanta/deploy/state.py

```diff
@@ -125,6 +181,73 @@ def reset(self) -> None:
         self.resource_state.clear()
         self.types_per_agent.clear()
 
+    def block_resource(self, resource: ResourceIdStr, details: ResourceDetails, transient: bool) -> None:
+        """
+        Mark the given resource as blocked, i.e. it's not deployable.
+
+        :param resource: The resource that should be blocked.
+        :param details: The details of the resource that should be blocked.
+        :param transient: True iff the given resource is blocked transitively. It's blocked because one of its dependencies
+                          is blocked.
+        """
+        self.resources[resource] = details
+        blocked_status = BlockedStatus.TRANSIENT if transient else BlockedStatus.YES
+        if resource in self.resource_state:
+            if not transient:
+                self.resource_state[resource].status = ResourceStatus.UNDEFINED
```

Comment:
Note to self: TODO: should this also set `HAS_UPDATE`? What about the dirty set?